airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/components/debezium/DebeziumComponent.kt

```diff
@@ -0,0 +1,389 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+package io.airbyte.cdk.components.debezium
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.NullNode
+import io.debezium.relational.history.HistoryRecord
+import java.time.Duration
+import java.util.*
+import kotlin.NoSuchElementException
+import org.apache.commons.math3.stat.descriptive.DescriptiveStatistics
+import org.apache.kafka.connect.source.SourceRecord
+
+/** API for collecting CDC records using Debezium. */
+fun interface DebeziumComponent {
+
+    /**
+     * Collect CDC records from the source database as specified by the [input]. This should have
+     * little to no side effects: the state of source database and the [input] determine the output.
+     *
+     * This is intended to be an atomic operation which should be straightforward to mock in
+     * high-level unit tests of a database source. As such its space and time usage are capped
+     * according to paramenters in the [input].
+     *
+     * In production, we need to checkpoint overall progress at regular intervals; the intention
+     * here is to do so after each call to [collect]. The [collectRepeatedly] method can be used to
+     * call [collect] in sequence.
+     */
+    fun collect(input: Input): Output
+
+    /**
+     * Input for [collect].
+     *
+     * This is separated into two parts according to the following principle:
+     * - [config] contains the input parameters which don't change from one successive call to
+     * [collect] to the next;
+     * - [state] contains those which do, these will be the initial content of Debezium's offset and
+     * schema history files.
+     */
+    @JvmRecord data class Input(val config: Config, val state: State)
+
+    /**
+     * [Config] holds the [Input] parameters which don't change from one successive call to
+     * [collect] to the next.
+     *
+     * The [debeziumProperties] property map is used to initialize the Debezium engine.
+     *
+     * The [lsnMapper] interface is used to compare relative CDC progress without knowledge of how
+     * the progress itself is encoded.
+     *
+     * The rest are termination criteria:
+     * - [upperBound] is the targeted upper bound of CDC progress, [collect] stops when it's reached
+     * or exceeded,
+     * - [maxRecords] is the threshold of collected [Record]s at or beyond which [collect] stops,
+     * - [maxRecordBytes] is the threshold of collected JSON [Record] bytes at or beyond which
+     * [collect] stops,
+     * - [maxTime] is the threshold of collected JSON-serialized [Record] bytes at or beyond which
+     * [collect] stops,
+     */
+    data class Config(
```

Comment:
You're right! Right now I have these objects defined as inner classes in the same file but the idea is to define some `api` subpackage where they wouldn't be inner classes. Because they're inner classes, I kept the names short, but this is just for ease of readability in the draft. The intention is to translate a name like `DebeziumComponent.Config` to `DebeziumComponentConfig` and do the same for all other API objects.